packages/itwin/tree-widget/src/components/trees/experimental/models-tree/ModelsVisibilityHandler.ts

```diff
@@ -0,0 +1,560 @@
+
+    const currNeverDrawn = new Set(this._viewport.neverDrawn ? this._viewport.neverDrawn : []);
+    const currAlwaysDrawn = new Set(this._viewport.alwaysDrawn ? this._viewport.alwaysDrawn : /* istanbul ignore next */ []);
+
+    if (on) {
+      changedNeverDraw = changedNeverDraw || currNeverDrawn.delete(elementId);
+      if ((!isDisplayedByDefault || isHiddenDueToExclusiveAlwaysDrawnElements) && !currAlwaysDrawn.has(elementId)) {
+        currAlwaysDrawn.add(elementId);
+        changedAlwaysDrawn = true;
+      }
+    } else {
+      changedAlwaysDrawn = changedAlwaysDrawn || currAlwaysDrawn.delete(elementId);
+      if (isDisplayedByDefault && !isHiddenDueToExclusiveAlwaysDrawnElements && !currNeverDrawn.has(elementId)) {
+        currNeverDrawn.add(elementId);
+        changedNeverDraw = true;
+      }
+    }
+
+    changedNeverDraw && this._viewport.setNeverDrawn(currNeverDrawn);
+    changedAlwaysDrawn && this._viewport.setAlwaysDrawn(currAlwaysDrawn, this._viewport.isAlwaysDrawnExclusive);
+  }
+
+  private isElementDisplayedByDefault(modelId: Id64String | undefined, categoryId: Id64String | undefined) {
+    return (
+      !!modelId &&
+      this.getModelDisplayStatus(modelId).state === "visible" &&
+      !!categoryId &&
+      this.getCategoryDisplayStatus(categoryId, modelId).state === "visible"
+    );
+  }
+
+  private hasExclusiveAlwaysDrawnElements() {
+    return this._viewport.isAlwaysDrawnExclusive && 0 !== this._viewport.alwaysDrawn?.size;
+  }
+
+  private onVisibilityChangeInternal() {
+    if (this._pendingVisibilityChange) {
+      return;
+    }
+
+    this._pendingVisibilityChange = setTimeout(() => {
+      this.onVisibilityChange.raiseEvent();
+      this._pendingVisibilityChange = undefined;
+    }, 0);
+  }
+
+  private getCategoryParentModelId(categoryNode: HierarchyNode): Id64String | undefined {
+    return categoryNode.extendedData && categoryNode.extendedData.modelIds[0] ? categoryNode.extendedData.modelIds[0][0] : /* istanbul ignore next */ undefined;
+  }
+
+  private getElementModelId(elementNode: HierarchyNode): Id64String | undefined {
+    return elementNode.extendedData ? elementNode.extendedData.modelId : /* istanbul ignore next */ undefined;
+  }
+
+  private getElementCategoryId(elementNode: HierarchyNode): Id64String | undefined {
+    return elementNode.extendedData ? elementNode.extendedData.categoryId : /* istanbul ignore next */ undefined;
+  }
+
+  private async getSubjectModelIds(subjectIds: Id64String[]) {
+    return (await Promise.all(subjectIds.map(async (id) => this._subjectModelIdsCache.getSubjectModelIds(id)))).reduce(
+      (allModelIds: Id64String[], curr: Id64String[]) => [...allModelIds, ...curr],
+      [],
+    );
+  }
+
+  private getAssemblyElementIds(assemblyId: Id64String) {
+    return this._elementIdsCache.getAssemblyElementIds(assemblyId);
+  }
+
+  // istanbul ignore next
+  private getGroupedElementIds(node: GroupingHierarchyNode & { key: GroupingNodeKey }) {
```

Comment:
is the ` & { key: GroupingNodeKey }` part needed?

Reply:
Removed